Implement software encoding (FFMPEG) for Windows

Ev1lbl0w opened the pull request:
![alvr](https://user-images.githubusercontent.com/6501975/154043919-b65780d5-dbb2-4ce6-acdb-84985c60cd4a.png)

This implements a software encoding backend *(using FFMPEG 5.0)* for Windows. Essentially I've copied the code from Linux implementation, and adapted it to work under Windows/DirectX.

New libraries are now bundled with the project, which came from [`ffmpeg-n5.0-latest-win64-gpl-5.0.zip`](https://github.com/BtbN/FFmpeg-Builds/releases)

The implementation isn't perfect, for example it doesn't skip encoded packets when too many are queued. This can happen when the headset is removed temporarily and put back on: the video feed will lag a few seconds behind while it catches up. I will be using this and testing some VR content on it now and hopefully fix any more woes that come up.

korejan commented:
If the sw encoder supports it can you add the option for 10-bit hevc, all you need to do is when the alvr option is on, switch the pixel format to AV_PIX_FMT_P010 (LE) and change the the codec profile to FF_PROFILE_HEVC_MAIN_10.

Couple of suggesttions. Try to not re-create the stagging texture everyframe, this only needs to be done when the size changes which shouldn't happen much throughtout a session when SteamVR dynamic resolution is on, default and not at all when user sets a fixed resolution (until they manually change it again). The last thing is a minor one, you can use Microsoft::WRL::ComPtr for RAII on directx interfaces so you don't have add Release everywhere or remember to increment the ref count.

korejan commented:
Maybe add an option to control the number of threads it can use, [AVCodecContext::thread_count](https://www.ffmpeg.org/doxygen/trunk/structAVCodecContext.html#aa852b6227d0778b62e9cc4034ad3720c)

zarik5 commented:
I think including all ffmpeg headers and binaries in the ALVR git repo is a but too much. You can set up xtask to download ffmpeg. Also you could setup a flag in xtask and a cargo feature in alvr_server to conditionally include ffmpeg.

Ev1lbl0w commented:
Thanks for all the comments, I've implemented all the suggested changes.

> switch the pixel format to AV_PIX_FMT_P010 (LE) and change the the codec profile to FF_PROFILE_HEVC_MAIN_10.

Regarding this, I've implemented this option for both h264 and hevc/h265 as the lib supports it. However, I wasn't able to get any output from the headset under this new setting, so this has to be fixed properly. h265 8-bit encoding in my setup does basically the same as #964; I had hopes it was just an issue on Linux, but apparently it's not.

> Maybe add an option to control the number of threads it can use, [AVCodecContext::thread_count](https://www.ffmpeg.org/doxygen/trunk/structAVCodecContext.html#aa852b6227d0778b62e9cc4034ad3720c)

Implemented with a new UI setting. Defaults to 0, which lets FFMPEG use as many as it sees fit 
![imagem](https://user-images.githubusercontent.com/6501975/154327615-21c1fe6a-e3fe-4553-8a35-8d9394ab6b28.png)

I've added other important, general info on the original post above.

xytovl commented:
hevc through x265 never worked, I tried to tweak the settings in any way that I could but without any success.

I did some tests with https://github.com/ultravideo/kvazaar that produced an image, but that wasn't very smooth, maybe you can have a look and tweak it better.

ColdIce1605 commented:
> Thanks for all the comments, I've implemented all the suggested changes.
> 
> > switch the pixel format to AV_PIX_FMT_P010 (LE) and change the the codec profile to FF_PROFILE_HEVC_MAIN_10.
> 
> Regarding this, I've implemented this option for both h264 and hevc/h265 as the lib supports it. However, I wasn't able to get any output from the headset under this new setting, so this has to be fixed properly. h265 8-bit encoding in my setup does basically the same as #964; I had hopes it was just an issue on Linux, but apparently it's not.
> 
> > Maybe add an option to control the number of threads it can use, [AVCodecContext::thread_count](https://www.ffmpeg.org/doxygen/trunk/structAVCodecContext.html#aa852b6227d0778b62e9cc4034ad3720c)
> 
> Implemented with a new UI setting. Defaults to 0, which lets FFMPEG use as many as it sees fit 
> ![imagem](https://user-images.githubusercontent.com/6501975/154327615-21c1fe6a-e3fe-4553-8a35-8d9394ab6b28.png)
> 
> I've added other important, general info on the original post above.
> 

This should probably also be added to linux

Ev1lbl0w commented:
> This should probably also be added to linux

Good point, it's also very simple to port so I'll do it right away.

Ev1lbl0w commented:
Ported 10-bit and thread_count changes to Linux, and fix some formatting issues with CI.

ColdIce1605 commented:
> Ported 10-bit and thread_count changes to Linux, and fix some formatting issues with CI.

Awesome, guessing it's only vaapi or software tho?

Ev1lbl0w commented:
> Awesome, guessing it's only vaapi or software tho?

Yup, just implemented on `EncodePipelineSW.cpp`, I can't test on hardware-accelerated versions.

ColdIce1605 commented:
Got it.

zarik5 commented on alvr/session/src/settings.rs:
```diff
@@ -178,6 +178,8 @@ pub struct VideoDesc {
 
     pub use_10bit_encoder: bool,
 
+    pub sw_thread_count: u32,
```
I think this should be advanced

```suggestion
    #[schema(advanced)]
    pub sw_thread_count: u32,
```

Ev1lbl0w replied:
In that case I was thinking if moving it here would make more sense? What do you think, is it ok as is or should I move it?
![imagem](https://user-images.githubusercontent.com/6501975/154717514-e4e444b8-87a8-4b0c-8c58-3bd3d655787a.png)

zarik5 reviewed (changes requested):
LGTM, just one small nit.
